### PR TITLE
Add new trace shard format

### PR DESCRIPTION
**Pull request by hikettei**
### Summary

- [x] add a new sharded safetensors trace format for speculator collect-traces
- [x] save raw logits (topk) and hidden states for generated tokens
- [ ] Delete `LalamoCompletion` and clean up diff

```
❯  git grep "LalamoCompletion"          
lalamo/data/lalamo_completions.py:class LalamoCompletion:
```

### CODE REVIEW

File: lalamo/data/trace_shard.py [hunk 1]
```diff
@@ -0,0 +1,219 @@
...
+    completion_token_ids: list[int]
+    raw_topk_token_ids: Int[Array, "completion_tokens k"]
+    raw_topk_token_logits: Float[Array, "completion_tokens k"]
+    output_norm_hidden: Float[Array, "completion_tokens hidden"]
+    layer_indices: tuple[int, ...] = tuple()
+    layer_hidden_states: Float[Array, "traced_layers completion_tokens hidden"] | None = None
+
+    def __post_init__(self) -> None:
+        completion_tokens = len(self.completion_token_ids)
+        if self.raw_topk_token_ids.shape[0] != completion_tokens:
```
Comment (knyazer) [inline lalamo/data/trace_shard.py:71]: use unpacking instead of indexing into shapes

File: lalamo/main.py [hunk 2]
```diff
@@ -963,14 +963,18 @@ def collect_traces(
...
             metavar="OUTPUT_PATH",
         ),
     ],
-    num_logits_per_token: Annotated[
+    raw_top_k: Annotated[
         int,
-        Option(help="Record logits for this number of most probable tokens"),
-    ] = 8,
+        Option(help="Record raw logits for this number of most probable tokens"),
+    ] = 8192,
```
Comment (knyazer) [inline lalamo/main.py:973]: why rename num_logits_per_token into raw_top_k? don't they have the same meaning?

File: lalamo/main.py [hunk 3]
```diff
@@ -990,16 +994,22 @@ def collect_traces(
             show_default="all",
         ),
     ] = None,
+    target_shard_size_mib: Annotated[
+        int,
+        Option(help="Approximate shard size budget in MiB before flushing a new safetensors shard"),
+    ] = DEFAULT_TRACE_SHARD_SIZE_BYTES // (1024 * 1024),
```
Comment (knyazer) [inline lalamo/main.py:1000]: this global constant is a not-so-important implementation detail; try to inline it as a default in the implementation of the actual writer and here set to some meaningful value that you won't have to go through mental gymnastics to compute

just not a big fan of a global constant that is used in only two places

File: lalamo/models/language_model.py [hunk 4]
```diff
@@ -347,7 +435,37 @@ def pad_and_repeat_state() -> tuple[DecodingState, GenerationStepResults]:
...
+            ),
+            (
+                rearrange(
+                    generated.layer_hidden_states,
+                    "iteration batch traced_layers hidden -> batch iteration traced_layers hidden",
+                )
+                if traced_layers
+                else None
+            ),
+        )
```
Comment (knyazer) [inline lalamo/models/language_model.py:468]: this is no good, i think; can't you have GenerationResults.init or something that would do all of this rearranging for you? especially because this whole thing is duplicated three times

File: lalamo/models/language_model.py [hunk 5]
```diff
@@ -307,17 +340,48 @@ def sample_and_update() -> tuple[DecodingState, GenerationStepResults]:
...
                 )
-                return new_state, GenerationStepResults(next_token_ids, next_top_k_token_ids, next_top_k_token_logits)
+                return new_state, GenerationStepResults(
+                    next_token_ids,
+                    next_top_k_token_ids,
+                    next_top_k_token_logits,
+                    (
+                        (raw_top_k := jax.lax.top_k(upcasted_logits, trace_config.raw_top_k))[1]
+                        if trace_config.raw_top_k is not None
+                        else None
```
Comment (knyazer) [inline lalamo/models/language_model.py:363]: why raw_top_k := blah?

also, just fyi, i don't think top_k works under sharding, make sure to check please (or just avoid it all together, since if you do 8000 top k its gonna be slower than sorting anyway)

File: lalamo/models/language_model.py [hunk 6]
```diff
@@ -74,12 +75,40 @@ class GenerationStepResults(NamedTuple):
...
 
 
 class GenerationResults(NamedTuple):
     token_ids: Int[Array, "batch response_tokens"]
     top_k_token_ids: Int[Array, "batch response_tokens k"] | None
     top_k_token_logits: Float[Array, "batch response_tokens k"] | None
+    raw_top_k_token_ids: Int[Array, "batch response_tokens raw_k"] | None
+    raw_top_k_token_logits: Float[Array, "batch response_tokens raw_k"] | None
+    output_norm_hidden: Float[Array, "batch response_tokens hidden"] | None
+    layer_hidden_states: Float[Array, "batch response_tokens traced_layers hidden"] | None
```
Comment (knyazer) [inline lalamo/models/language_model.py:91]: what's the difference between top_k_token_ids and raw_top_k_token_ids?

File: lalamo/data/trace_shard.py [hunk 7]
```diff
@@ -0,0 +1,219 @@
...
+    @classmethod
+    def from_path(cls, path: Path) -> "TraceShard":
+        with path.open("rb") as fd:
+            _, tensors = safe_read(fd)
+            loaded_tensors = {
+                "prefix_offsets": tensors["prefix_offsets"],
+                "prefix_token_ids": tensors["prefix_token_ids"],
+                "completion_offsets": tensors["completion_offsets"],
+                "completion_token_ids": tensors["completion_token_ids"],
+                **{field_name: tensors[field_name] for field_name in _PER_COMPLETION_TOKEN_FIELDS},
```
Comment (knyazer) [inline lalamo/data/trace_shard.py:119]: i feel like most of this file is genuinely unnecessary, you've introduced lots of abstraction with very tricky semantics - like this unpacking is kinda insane, you can get a lot of silent bugs here by overriding the "something something": tenstors["something something"] on accident. and you have this in several places. I think you can get away with just TraceCompletionRecord - we do need to find a better name fyi, maybe CompletionWithHidden or something, trace is a very overloaded term - and that's it. just like, remove all this complexity, write a few functional-style pure helpers that dump/load the trace into safetensors, and probably this would work?

I haven't read your implementation super thoroughly to actually understand whether this will work or not, but I think you can viably get to ~150 lines in total for this PR

File: lalamo/data/trace_shard.py [hunk 8]
```diff
@@ -0,0 +1,219 @@
...
+            )
+
+
+@dataclass
+class TraceShardWriter:
+    output_dir: Path
+    target_shard_size_bytes: int = DEFAULT_TRACE_SHARD_SIZE_BYTES
+    shard_index: int = 0
+    _records: list[TraceCompletionRecord] = field(default_factory=list, repr=False)
+    _estimated_size_bytes: int = 0
```
Comment (knyazer) [inline lalamo/data/trace_shard.py:171]: why do you store this as a property that is also private - why not recompute it on the fly?

**Comment by hikettei** [comment]
@knyazer thanks for the review, but i will reopen this pr with ~+150 lines approach by replacing LalamoCompletion
